File: frontend/src/engine/capabilities/freezePanes.ts
/**
 * freezePanes – Freeze rows and/or columns.
 *
 * Office.js notes:
 * - WorksheetFreezePanes API is in ExcelApi 1.7+.
 * - freezeAt(range) freezes rows above and columns left of the given range.
 * - e.g. freezeAt("B2") freezes row 1 and column A.
 */

import { CapabilityMeta, FreezePanesParams, StepResult, ExecutionOptions } from "../types";
import { registry } from "../capabilityRegistry";

const meta: CapabilityMeta = {
  action: "freezePanes",
  description: "Freeze rows and/or columns at a specified cell",
  mutates: false,
  affectsFormatting: false,
  requiresApiSet: "ExcelApi 1.7",
};

async function handler(
  context: Excel.RequestContext,
  params: FreezePanesParams,
  options: ExecutionOptions
): Promise<StepResult> {
  const { cell, sheetName } = params;

  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: `Would freeze panes at ${cell}`,
    };
  }

  options.onProgress?.(`Freezing panes at ${cell}...`);

  // Validate and resolve target sheet
  let sheet: Excel.Worksheet;
  if (sheetName) {
    const ws = context.workbook.worksheets.getItemOrNullObject(sheetName);
    ws.load("isNullObject");
    await context.sync();
    if (ws.isNullObject) {
      return {
        stepId: "",
        status: "error",
        message: `Sheet "${sheetName}" not found. Please check the sheet name.`,
      };
    }
    sheet = ws;
  } else {
    sheet = context.workbook.worksheets.getActiveWorksheet();
  }

  const range = sheet.getRange(cell);
  sheet.freezePanes.freezeAt(range);
  await context.sync();

  return {
    stepId: "",
    status: "success",
    message: `Froze panes at ${cell}${sheetName ? ` on "${sheetName}"` : ""}`,
  };
}

// ── Legacy-Excel fallback (ExcelApi < 1.7) ────────────────────────────────────
// WorksheetFreezePanes requires 1.7. Before that, Office.js offers no
// programmatic freeze. We can't simulate the split-scrolling behavior
// through any other primitive, so we gracefully skip with a warning — the
// user can freeze manually via View › Freeze Panes. Status=success because
// this is cosmetic and shouldn't abort a data plan.
async function fallback(
  _context: Excel.RequestContext,
  params: FreezePanesParams,
  options: ExecutionOptions,
): Promise<StepResult> {
  if (options.dryRun) {
    return {
      stepId: "",
      status: "success",
      message: `Would skip freeze at ${params.cell} (legacy fallback; requires ExcelApi 1.7+).`,
    };
  }

  options.onProgress?.("Legacy-Excel mode: freeze panes unavailable, skipping...");

  return {
    stepId: "",
    status: "success",
    message:
      `Freeze panes at ${params.cell} skipped — WorksheetFreezePanes requires ExcelApi 1.7+ ` +
      `(Excel 2019 / 2021 / Microsoft 365). Use View › Freeze Panes manually ` +
      `(legacy-Excel fallback).`,
  };
}

registry.register(meta, handler as any, fallback as any);
export { meta };

File: backend/app/execution/capabilities/freeze_panes.py
"""freezePanes — freeze the top rows / left columns of a sheet."""

from __future__ import annotations

from typing import Any

from app.execution.base import ExecutorContext
from app.execution.capability_registry import registry


def handler(ctx: ExecutorContext, params: dict[str, Any]) -> dict[str, Any]:
    rows = int(params.get("rowCount", 0))
    cols = int(params.get("columnCount", 0))
    sheet_name = params.get("sheetName")

    book = ctx.workbook_handle
    sheet = book.sheets[sheet_name] if sheet_name else book.sheets.active

    if ctx.dry_run:
        return {
            "status": "preview",
            "message": f"Would freeze {rows} row(s) and {cols} column(s) on {sheet.name}.",
        }

    # Freeze panes requires activating the sheet and then setting SplitRow/SplitColumn
    # on the active window. xlwings gives us the underlying Window via book.app.api.
    try:
        sheet.activate()
        app_api = book.app.api
        window = app_api.ActiveWindow
        # Clear any existing freeze first.
        window.FreezePanes = False
        window.SplitRow = rows
        window.SplitColumn = cols
        window.FreezePanes = True
    except Exception as exc:  # noqa: BLE001
        return {"status": "error", "message": f"Freeze panes failed: {exc}", "error": str(exc)}

    return {
        "status": "success",
        "message": f"Froze {rows} row(s) / {cols} column(s) on {sheet.name}.",
        "outputs": {"sheet": sheet.name, "rowCount": rows, "columnCount": cols},
    }


registry.register("freezePanes", handler, mutates=False, affects_formatting=True)
